File: macos/Sources/TethertoneMac/Pairing/TransportRoute.swift
import Foundation

/// Mirrors `TransportRoute` in the Kotlin shared module — both sides must
/// classify an address the same way or they will describe the same connection
/// differently.
enum TransportRoute: String, CaseIterable, Identifiable {
    case usb, lan, vpn, internet

    var id: String { rawValue }

    var label: String {
        switch self {
        case .usb: return "USB cable"
        case .lan: return "Same Wi-Fi"
        case .vpn: return "VPN"
        case .internet: return "Internet"
        }
    }

    var detail: String {
        switch self {
        case .usb: return "Lowest latency, no mobile data, never drops."
        case .lan: return "Both devices on one network."
        case .vpn: return "Works anywhere, including mobile data."
        case .internet: return "Through a tunnel or a forwarded port."
        }
    }

    var symbol: String {
        switch self {
        case .usb: return "cable.connector"
        case .lan: return "wifi"
        case .vpn: return "lock.shield"
        case .internet: return "globe"
        }
    }

    /// Shown when the route is not on offer — the one thing to do about it.
    var missingAdvice: String {
        switch self {
        case .usb: return "Plug the phone in with USB debugging enabled."
        case .lan: return "No local address — check this Mac is on a network."
        case .vpn: return "Install Tailscale on both devices; its address is picked up automatically."
        case .internet: return "Add a tunnel hostname under Network."
        }
    }
}

/// `127.0.0.1` means the cable here: loopback only appears in a pairing code
/// after `adb reverse` has made the phone's own localhost reach this Mac.
func classifyHost(_ host: String) -> TransportRoute {
    let trimmed = host.trimmingCharacters(in: .whitespaces).lowercased()
    if trimmed == "localhost" || trimmed.hasPrefix("127.") { return .usb }

    let octets = trimmed.split(separator: ".").compactMap { Int($0) }
    guard octets.count == 4, octets.allSatisfy({ (0...255).contains($0) }) else {
        return .internet      // a hostname, so it was typed in or came from a tunnel
    }
    switch (octets[0], octets[1]) {
    case (100, 64...127): return .vpn
    case (10, _): return .lan
    case (192, 168): return .lan
    case (172, 16...31): return .lan
    case (169, 254): return .lan
    default: return .internet
    }
}

File: android/shared/src/commonMain/kotlin/dev/audiobridge/shared/TransportRoute.kt
package dev.audiobridge.shared

/**
 * How a sink is reaching, or could reach, the server.
 *
 * The three routes behave very differently — a cable costs nothing and never
 * drops, Wi-Fi needs both devices on one network, and anything beyond that
 * needs a VPN or tunnel because neither a home router nor a mobile carrier will
 * accept an inbound connection. Telling them apart is what lets both apps say
 * something useful instead of "connecting…".
 */
enum class TransportRoute {
    USB,
    LAN,
    VPN,
    INTERNET;

    val label: String
        get() = when (this) {
            USB -> "USB cable"
            LAN -> "Same Wi-Fi"
            VPN -> "VPN"
            INTERNET -> "Internet"
        }

    val detail: String
        get() = when (this) {
            USB -> "Lowest latency, no data used, never drops."
            LAN -> "Both devices on one network."
            VPN -> "Works anywhere, including mobile data."
            INTERNET -> "Through a tunnel or forwarded port."
        }

    /** Uses mobile data, so bandwidth is worth mentioning. */
    val costsMobileData: Boolean
        get() = this == VPN || this == INTERNET
}

/**
 * `127.0.0.1` means the USB tunnel here: the server only ever puts loopback in
 * a pairing code after `adb reverse` has made the phone's own localhost reach
 * the Mac, so from the phone's side it is the cable.
 */
fun classifyHost(host: String): TransportRoute {
    val trimmed = host.trim().lowercase()
    if (trimmed == "localhost" || trimmed.startsWith("127.")) return TransportRoute.USB

    val octets = trimmed.split(".").mapNotNull { it.toIntOrNull() }
    if (octets.size != 4 || octets.any { it !in 0..255 }) {
        // A hostname, so it was typed in or came from a tunnel.
        return TransportRoute.INTERNET
    }
    val (a, b) = octets[0] to octets[1]
    return when {
        // Tailscale, ZeroTier and friends hand out carrier-grade NAT space.
        a == 100 && b in 64..127 -> TransportRoute.VPN
        a == 10 -> TransportRoute.LAN
        a == 192 && b == 168 -> TransportRoute.LAN
        a == 172 && b in 16..31 -> TransportRoute.LAN
        a == 169 && b == 254 -> TransportRoute.LAN
        else -> TransportRoute.INTERNET
    }
}

/** The distinct routes a pairing code offers, in the order they will be tried. */
fun PairingInfo.routes(): List<TransportRoute> =
    hosts.map(::classifyHost).distinct()
